pkg/controller/direct/compute/forwardingrule_controller.go

```diff
@@ -0,0 +1,460 @@
+	// Use setTarget and setLabels to update target and labels fields.
+	var err error
+	// TODO(yuhou): Checked the realGCP logs, setLabels request is being sent even when there are no updates to labels.
+	// That might because of the generated labelsFingerPrint?
+	if a.Location == "global" {
+		setLabelsReq := &computepb.SetLabelsGlobalForwardingRuleRequest{
+			Resource:                       a.resourceID,
+			GlobalSetLabelsRequestResource: &computepb.GlobalSetLabelsRequest{LabelFingerprint: a.actual.LabelFingerprint, Labels: forwardingRule.Labels},
+			Project:                        a.projectID,
+		}
+		_, err = a.globalForwardingRulesClient.SetLabels(ctx, setLabelsReq)
+	} else {
+		setLabelsReq := &computepb.SetLabelsForwardingRuleRequest{
+			Resource:                       a.resourceID,
+			RegionSetLabelsRequestResource: &computepb.RegionSetLabelsRequest{LabelFingerprint: a.actual.LabelFingerprint, Labels: forwardingRule.Labels},
+			Project:                        a.projectID,
+			Region:                         a.Location,
+		}
+		_, err = a.forwardingRulesClient.SetLabels(ctx, setLabelsReq)
+	}
+	if err != nil {
+		return fmt.Errorf("updating ComputeForwardingRule labels %s failed: %w", a.resourceID, err)
+	}
+	log.V(2).Info("successfully updated ComputeForwardingRule labels", "name", a.resourceID)
+
+	// setTarget request is sent when there are updates to target.
+	if !reflect.DeepEqual(forwardingRule.Target, a.actual.Target) {
+		if a.Location == "global" {
+			setTargetReq := &computepb.SetTargetGlobalForwardingRuleRequest{
+				ForwardingRule:          a.resourceID,
+				TargetReferenceResource: &computepb.TargetReference{Target: forwardingRule.Target},
+				Project:                 a.projectID,
+			}
+			_, err = a.globalForwardingRulesClient.SetTarget(ctx, setTargetReq)
+		} else {
+			setTargetReq := &computepb.SetTargetForwardingRuleRequest{
+				ForwardingRule:          a.resourceID,
+				TargetReferenceResource: &computepb.TargetReference{Target: forwardingRule.Target},
+				Project:                 a.projectID,
+				Region:                  a.Location,
+			}
+			_, err = a.forwardingRulesClient.SetTarget(ctx, setTargetReq)
+		}
+		if err != nil {
+			return fmt.Errorf("updating ComputeForwardingRule target %s failed: %w", a.resourceID, err)
+		}
+		log.V(2).Info("successfully updated ComputeForwardingRule target", "name", a.resourceID)
+	}
+
+	updated := &computepb.ForwardingRule{}
+	if a.Location == "global" {
+		getReq := &computepb.GetGlobalForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Project:        a.projectID,
+		}
+		updated, err = a.globalForwardingRulesClient.Get(ctx, getReq)
+	} else {
+		getReq := &computepb.GetForwardingRuleRequest{
+			ForwardingRule: a.resourceID,
+			Region:         a.Location,
+			Project:        a.projectID,
+		}
+		updated, err = a.forwardingRulesClient.Get(ctx, getReq)
+	}
+	if err != nil {
+		if direct.IsNotFound(err) {
```

Comment:
Not sure you need the IsNotFound check here, this is pretty surprising.

Reply:
Removed. If update succeeds, the resource must exist.